cylc/flow/data_store_mgr.py

```diff
@@ -1524,7 +1526,7 @@ def update_data_structure(self, reloaded=False):
             # Gather this batch of deltas for publish
             self.publish_deltas = self.get_publish_deltas()
 
-        self.updates_pending = next_update_pending
+        self.updates_pending = self.prune_pending
```

Comment:
```suggestion
        self.updates_pending = False
```

Reply:
Not quite, the extra flag actually provides an additional call to `update_data_structure` after pruning (which is needed because pruning effects state totals).

Reply:
I now see `prune_data_store()` may set `self.prune_pending` to `True`. But it also sets `self.updates_pending` to `True`, which leads me to believe this line can be removed altogether.

```suggestion
```

Reply:
I don't think so, as that's what flags `update_data_structure` to run again (it's checked in the scheduler main loop)

Reply:
It also sets `self.updates_pending`  to true for the `if` block below `prune_data_store()`  ... Probably a better way of doing it, but the justification is there.

Reply:
Essentially when `update_data_structure` is called .. If pruning was done, not only run all the other updating (including delta application and publishing), but run the `update_data_structure` again (after pruning has been applied) to recalculate state totals etc for application and publishing.

Reply:
If we apply the pruning beforehand, we may loose intermediate states.

Reply:
Also, `update_data_structure` can be called without `self.updates_pending`  otherwise we could get rid of that `if self.updates_pending` block  and set it to `False` at the top .. 
If we broke up `update_data_structure` into two separate functions, maybe, but what if there are also updates pending for the alternate call?

Reply:
I've renamed it `self.updates_pending_follow_on`  .. Just for clarity.

Reply:
Sorry but unless I'm missing something, `self.updates_pending_follow_on` is not actually used. It already matches the value of `self.updates_pending` here in `prune_data_store()`
https://github.com/cylc/cylc-flow/blob/54f4756a4c3fe80dd8edc8a95f5ca78878274896/cylc/flow/data_store_mgr.py#L1597-L1598

If you are saying that the values of `self.updates_pending_follow_on` and `self.updates_pending` can diverge in between here
https://github.com/cylc/cylc-flow/blob/54f4756a4c3fe80dd8edc8a95f5ca78878274896/cylc/flow/data_store_mgr.py#L1503-L1504

and this line that this comment thread is on
https://github.com/cylc/cylc-flow/blob/54f4756a4c3fe80dd8edc8a95f5ca78878274896/cylc/flow/data_store_mgr.py#L1529

?

Reply:
Ah I guess if `self.updates_pending` is `True` to begin with it may not get set to `False`.

I think this would simplify it a bit
```diff
diff --git a/cylc/flow/data_store_mgr.py b/cylc/flow/data_store_mgr.py
index a15341d8f..982fe046d 100644
--- a/cylc/flow/data_store_mgr.py
+++ b/cylc/flow/data_store_mgr.py
@@ -509,5 +509,4 @@ class DataStoreMgr:
         self.pruned_task_proxies = set()
         self.updates_pending = False
-        self.updates_pending_follow_on = False
         self.publish_pending = False
 
@@ -1501,6 +1500,5 @@ class DataStoreMgr:
             self.next_n_edge_distance = None
 
-        self.updates_pending_follow_on = False
-        self.prune_data_store()
+        follow_on = self.prune_data_store()
         if self.updates_pending:
             # update
@@ -1527,5 +1525,5 @@ class DataStoreMgr:
             self.publish_deltas = self.get_publish_deltas()
 
-        self.updates_pending = self.updates_pending_follow_on
+        self.updates_pending = follow_on
 
         # Clear deltas
@@ -1533,11 +1531,14 @@ class DataStoreMgr:
         self.clear_delta_store()
 
-    def prune_data_store(self):
-        """Remove flagged nodes and edges not in the set of active paths."""
+    def prune_data_store(self) -> bool:
+        """Remove flagged nodes and edges not in the set of active paths.
+
+        Returns whether updates are pending after this.
+        """
 
         self.family_pruned_ids.clear()
 
         if not self.prune_flagged_nodes:
-            return
+            return False
 
         # Keep all nodes in the path of active tasks.
@@ -1595,6 +1596,7 @@ class DataStoreMgr:
         if node_ids:
             self.pruned_task_proxies.update(node_ids)
-            self.updates_pending = True
-            self.updates_pending_follow_on = True
+            return True
+
+        return False
 
     def _family_ascent_point_prune(
```

Reply:
No, it wouldn't .. If there was no pruning, then it would get set to `False`..
The reason I would keep it as a separate variable is, if another operation requires a follow-on update.

Reply:
> Unless you are saying that the values of `self.updates_pending_follow_on` and `self.updates_pending` can diverge in between here

Exactly, no pruning, they diverge.. All this:
```
        if node_ids:
            self.pruned_task_proxies.update(node_ids)
            self.updates_pending = True
            self.updates_pending_follow_on = True
```
Says is (the `if node_ids:` is key) if pruning was actually done, update all the other stuff (states ..etc)..
(note: running `self.prune_data_store()` doesn't imply it actually found something to prune, but if it did, do an update, and a follow on update)

Reply:
Ok I think I'm convinced now 😁